main/src/ca/uwaterloo/flix/language/phase/unification/Zhegalkin.scala

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2024 Magnus Madsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.language.phase.unification.set.SetFormula
+import ca.uwaterloo.flix.language.phase.unification.set.SetFormula.{Compl, Cst, ElemSet, Inter, Union, Var}
+import ca.uwaterloo.flix.util.CofiniteIntSet
+
+import scala.collection.immutable.SortedSet
+
+object Zhegalkin {
+
+  /** Represents a variable. */
+  case class ZhegalkinVar(v: Int) extends Ordered[ZhegalkinVar] {
+    override def toString: String = s"x$v"
+
+    override def compare(that: ZhegalkinVar): Int = this.v.compare(that.v)
+  }
+
+  /** Represents a set Zhegalkin constant (i.e. a set or co-set) */
+  case class ZhegalkinConstant(s: CofiniteIntSet) {
+    def compl: ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.complement(s))
+    def union(that: ZhegalkinConstant): ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.union(s, that.s))
+    def inter(that: ZhegalkinConstant): ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.intersection(s, that.s))
+
+    override def toString: String = {
+      if (s.isEmpty) "Ø"
+      else if (s.isUniverse) "T"
+      else s match {
+        case CofiniteIntSet.Set(xs) => s"{${xs.mkString(", ")}}"
+        case CofiniteIntSet.Compl(xs) => s"~{${xs.mkString(", ")}}"
+      }
+    }
+  }
+
+  private val ZEmpty: ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.empty)
+  private val ZUniverse: ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.universe)
+
+  /** Represents a Zhegalkin term: c n x1 n x2 n ... n xn */
+  case class ZhegalkinTerm(cst: ZhegalkinConstant, vars: SortedSet[ZhegalkinVar]) {
+    override def toString: String =
+      if (vars.isEmpty)
+        cst.toString
+      else
+        s"$cst n ${vars.mkString(" n ")}"
+  }
+
+  /** Represents a Zhegalkin expr: c ? t1 ? t2 ? ... ? tn */
+  case class ZhegalkinExpr(cst: ZhegalkinConstant, terms: List[ZhegalkinTerm]) {
+    override def toString: String =
+      if (terms.isEmpty)
+        cst.toString
+      else
+        s"$cst ? ${terms.map(t => s"($t)").mkString(" ? ")}"
+  }
+
+  /**
+    * A smart constructor to filter empty intersections.
```

Comment:
Do we have an invariant that intersections are never empty? If so, it should be noted.

Reply:
No, its just an optimization.